docs/multiwiki.md

```diff
@@ -0,0 +1,60 @@
+Background
+==========
+This application was originally written for English Wikipedia, and for
+convenience the primary IDs of many entities were kept in sync with MediaWiki
+database IDs.
+
+Multiwiki feature
+==================
+In February 2016, we've added support for articles from any of the [Wikimedia
+project wikis](https://meta.wikimedia.org/wiki/Wikimedia_wikis#Core_projects),
+which involves database schema and code changes.  The `Wiki` table will hold
+one record for each wiki referenced from the application, and all onwiki
+entities will link to that row.  The entity primary IDs will be decoupled from
+MediaWiki ID and new `mw_page_id` columns added to hold the MediaWiki primary
+ID.  All code will be rewritten to distinguish between cases where we want the
+MediaWiki ID (during sync and deduplication), or the local primary ID
+(concisely identifies local articles).
+
+The new schema looks like this:
+![multiwiki schema](./multiwiki.svg)
+
+We haven't finished designing how User and Course associate with wikis, so we
+use two shorthand concepts for the moment.  There is the list of wikis that each
+user and collectively, each course edits.  One wiki from this list is arbitrarily
+chosen as the home wiki, otherwise we fall back to the application default wiki.
+This is a pretty bad guess, and will be changed in the future.  Having a home
+wiki is helpful to minimize code changes during this refactor, but we don't
+know whether we even want such a property.
+
+Migration
+=========
+There will be a one-time schema migration, which is mostly reversible in case
+of rollback.  Since everything in the application database is on the default
+wiki, we can simply copy the primary IDs over to the MediaWiki ID columns, and
+set the wiki to our default record.
+
+As multiwiki entities are added to the system, the reverse migration will get
+riskier.  If two entities have the same MediaWiki primary ID, one will be lost
+during a rollback.  We should test for this condition manually if that becomes
+necessary, the migration itself doesn't include error checking.
+
+The refactor is split into two branches, the first, `wiki_schema`, creates
+a schema without making logic changes.  The second, `use_native_ids`, rewrites
+the application to distinguish between local article ID and MediaWiki page ID.
+That branch is organized with safe patches towards the beginning, and riskier
+ones towards the end, so that a partial rollback is possible if required.
+Partially rolled back code will be protected by the default value logic, and
+will create valid records, but will only handle English Wikipedia articles, so
+should be corrected and redeployed as quickly as possible.
+
```

Comment:
K I think I understand... Mainly I'd suggest that we limit possible rollback plans on the basis of what's actually risky and what's controllable given the current install base. It seems that a script copying existing primary keys and setting `wiki_id` is sure to succeed and can be easily verified by queries. This would happen before any of the code that actually uses the new columns is rolled out, right? So the default values set in `before_validation` in the models wouldn't be needed... Then when it's ready, multiwiki code can be rolled out on WikiEdu's production instance, but without providing any UI features that let you do anything anywhere other than a default wiki. WikiEdu I'm sure could guarantee they won't switch the default wiki during whatever testing period they need. That would guarantee no duplicate ids during that testing period. When they're satisfied with the stability of the new code, they could roll out UI for multiwiki features, if they wish (or maybe they'll want some switches to hide it, not sure...).

A similar plan could be enacted for the semi-production instance on labs.

For the real WMF P&E production instance, there's no worry about rollback, since it doesn't yet exist. :)

Does that make sense? Have I misunderstood something? BTW I'm not sure what you mean by, "Partially rolled back code will be protected by the default value logic, and will create valid records." Maybe the idea isn't to fill in the new columns in one go with a script? Apologies if I'm being thick-skulled!!